beacon_node/beacon_chain/src/block_verification.rs

```diff
@@ -520,16 +520,71 @@ impl<T: BeaconChainTypes> GossipVerifiedBlock<T> {
             &chain.store,
         )?;
 
-        let (mut parent, block) = load_parent(block, chain)?;
+        let block_epoch = block.slot().epoch(T::EthSpec::slots_per_epoch());
+        let (parent_block, block) = verify_parent_block_is_known(chain, block)?;
+
+        // Paranoid check to prevent propagation of blocks that don't form a legitimate chain.
+        //
+        // This is not in the spec, but @protolambda tells me that the majority of other clients are
+        // already doing it. For reference:
+        //
+        // https://github.com/ethereum/eth2.0-specs/pull/2196
+        if parent_block.slot >= block.slot() {
+            return Err(BlockError::BlockIsNotLaterThanParent {
+                block_slot: block.slot(),
+                parent_slot: parent_block.slot,
+            });
+        }
+
+        let proposer_shuffling_decision_block =
+            if parent_block.slot.epoch(T::EthSpec::slots_per_epoch()) == block_epoch {
+                parent_block
+                    .next_epoch_shuffling_id
```

Comment:
To clarify: this is exploiting the fact that the `next_epoch_shuffling_id` (from the PoV of the attester shuffling) is the last block root of `block_epoch - 1`, which is _also_ the shuffling ID for the proposer shuffling for `block_epoch`? Nice! :sunglasses:

Reply:
I was happy with this too!